File: 36-valid-sudoku/valid-sudoku.java
import java.util.*;

class Solution {
    public boolean isValidSudoku(char[][] board) {

        Set<Character>[] rows = new HashSet[9];
        Set<Character>[] cols = new HashSet[9];
        Set<Character>[] boxes = new HashSet[9];

        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<>();
            cols[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        for (int r = 0; r < 9; r++) {

            for (int c = 0; c < 9; c++) {

                char num = board[r][c];

                if (num == '.') {
                    continue;
                }

                int box = (r / 3) * 3 + (c / 3);

                if (
                    rows[r].contains(num)
                    || cols[c].contains(num)
                    || boxes[box].contains(num)
                ) {
                    return false;
                }

                rows[r].add(num);
                cols[c].add(num);
                boxes[box].add(num);
            }
        }

        return true;
    }
}

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board):
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for r in range(9):
            for c in range(9):

                num = board[r][c]

                if num == ".":
                    continue

                box = (r // 3) * 3 + (c // 3)

                if (
                    num in rows[r]
                    or num in cols[c]
                    or num in boxes[box]
                ):
                    return False

                rows[r].add(num)
                cols[c].add(num)
                boxes[box].add(num)

        return True